agent/agent-profiler/agent-alerting/src/main/java/com/microsoft/applicationinsights/alerting/analysis/aggregations/WindowedAggregation.java

```diff
@@ -19,62 +19,56 @@
  * DEALINGS IN THE SOFTWARE.
  */
 
-package com.microsoft.applicationinsights.alerting.analysis;
+package com.microsoft.applicationinsights.alerting.analysis.aggregations;
 
+import com.microsoft.applicationinsights.alerting.analysis.TimeSource;
+import com.microsoft.applicationinsights.alerting.analysis.data.TelemetryDataPoint;
 import java.time.Instant;
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.List;
-import java.util.OptionalDouble;
-import java.util.function.Consumer;
 
-/** Applies a time window to data and calculates a mean of the data during that window. */
-public class RollingAverage {
+/** Applies a time window to data. I.e holds the last n seconds of data. */
+public class WindowedAggregation {
 
   private final long windowLengthInSec;
   private final TimeSource timeSource;
-  private final List<TelemetryDataPoint> telemetryDataPoints = new ArrayList<>();
-  private Consumer<Double> consumer;
+  private final List<TelemetryDataPoint> telemetryDataPoints =
+      Collections.synchronizedList(new ArrayList<>());
   private static final int DEFAULT_ROLLING_AVERAGE_WINDOW_IN_SEC =
       Integer.parseInt(
           System.getProperty(
               "applicationinsights.preview.profiler.rolling-average-window-in-sec", "120"));
```

Comment:
Since we know the percentile (e.g. 95th) and the threshold (e.g. 5000 millis), maybe we can do something smart like keep 1-second buckets that store just the request count that exceeded the threshold and the total request count. In addition to taking less memory, should be a lot faster to calculate if the 95th percentile exceeded 5000 milliseconds compared to (whatever magic guava Quantiles does). Also maybe this is our way out of the guava dependency 🤔.